text/204-golang-bindings.md

```diff
@@ -0,0 +1,655 @@
+  Retries int
+  StartPeriod cdk.Duration;
+  Timeout cdk.Duration;
+}
+```
+
+This would allow a function that takes that interface as an argument, such as
+[renderHealthCheck](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-ecs/lib/container-definition.ts#L652-L660) look like this
+(simplified, without accounting for default values):
+
+```go
+func renderHealthCheck(hc HealthCheck) CfnTaskDefinition.HealthCheckProperty {
+    return CfnTaskDefintion.HealthCheckProperty{
+        command:     hc.Command,
+        interval:    hc.Interval,
+        retries:     hc.Retries,
+        startPeriod: hc.StartPeriod,
+        timeout:     hc.Timeout,
+    }
+}
+
+```
+
+Rather than:
+
+```go
+func renderHealthCheck(hc HealthCheck) CfnTaskDefinition.HealthCheckProperty {
+    return CfnTaskDefintion.HealthCheckProperty{
+        command:     hc.GetCommand(),
+        interval:    hc.GetInterval(),
+        retries:     hc.GetRetries(),
+        startPeriod: hc.GetStartPeriod(),
+        timeout:     hc.GetTimeout(),
+    }
+}
+```
+
+#### Questions
+
+* Are properties on datatype interfaces always [immutable](https://github.com/aws/jsii/blob/master/packages/%40jsii/spec/lib/assembly.ts#L607-L611)?
+  (All the datatype interfaces in the [JSII calc test assembly](https://github.com/aws/jsii/blob/master/packages/jsii-calc/test/assembly.jsii) appear
+to only have immutable, i.e. readonly, properties).
+
+## Classes
+
+Typescript takes an object-oriented approach to classes, which includes using polymorphism and subtyping, neither of which are natively supported in
+Go, which is not an object-oriented language. While custom structs, which can be used as pointer receivers in function definitions to simulate
+methods", can be used to encapsulate object-like behavior in Go, subtyping on these custom structs is not possible. In order to simulate subtyping, we
+would need to generate an interface in addition to a concrete struct for each JSII class.
+
+The JSII [ClassType](https://github.com/aws/jsii/blob/master/packages/%40jsii/spec/lib/assembly.ts#L803) assembly provides information on whether a
+class is abstract, whether it extends another class, and whether it implements other interfaces. We will discuss each case in turn.
+
+### Case 1: Simple class
+
+```ts
+class Greeter {
+    greeting: string;
+
+    constructor(message: string) {
+        this.greeting = message;
+    }
+
+    greet(): string {
+        return "Hello, " + this.greeting;
+    }
+}
+
+let greeter = new Greeter("world");
+greeter.greet() // "Hello, world"
+```
+
+In Go: ([Go playground example](https://play.golang.org/p/cMW3Rv0lERh))
+
+```go
+package greeter
+
+type GreeterIface interface {
+    Greet() string
+}
+
+type Greeter struct {
+    greeting string
+}
+
+func New(message string) *Greeter {
+    return &Greeter{message}
+}
+func (g *Greeter) Greet() string {
+    return fmt.Sprintf("Hello, %+s", g.greeting)
+}
+
+// usage
+g := greeter.New("world")
```

Comment:
Where is the public `greeting` property? It should also be part of the `GreeterIface`, no?

Reply:
I was trying to use a simplified example, so this was taken from https://www.typescriptlang.org/docs/handbook/classes.html#classes. I can replace this with an actual class from the CDK instead if that is preferred.

Reply:
Jsii has a huge set of tests in the "calc" suite. Eventually we will need to show how each one translates to go. Maybe that could be a good way to ensure coverage of this rfc? (At least as inspiration)